persistent_stream.go

```diff
@@ -45,67 +45,63 @@ func (d *Daemon) handlePersistentConn(r ggio.Reader, unsafeW ggio.WriteCloser) {
 			return
 		}
 
-		callID, err := uuid.FromBytes(req.CallId)
-		if err != nil {
-			log.Debugw("bad call id: ", "error", err)
-			continue
+		go d.handlePersistentConnRequest(req, w, &streamHandlers)
+	}
+}
+
+func (d *Daemon) handlePersistentConnRequest(req pb.PersistentConnectionRequest, w ggio.WriteCloser, streamHandlers *[]string) {
+	callID, err := uuid.FromBytes(req.CallId)
+	if err != nil {
+		log.Debugw("bad call id: ", "error", err)
+		return
+	}
+
+	switch req.Message.(type) {
+	case *pb.PersistentConnectionRequest_AddUnaryHandler:
+		resp := d.doAddUnaryHandler(w, callID, req.GetAddUnaryHandler())
+
+		d.mx.Lock()
+		if _, ok := resp.Message.(*pb.PersistentConnectionResponse_DaemonError); !ok {
+			*streamHandlers = append(
+				*streamHandlers,
+				*req.GetAddUnaryHandler().Proto,
+			)
```

Comment:
`append` is [not thread-safe](https://stackoverflow.com/questions/44152988/append-not-thread-safe), let's use a mutex.

Reply:
what's wrong with the mutex on line 63?

Reply:
Missed it, thanks :)